R/data.table.R

```diff
@@ -2291,25 +2291,10 @@ transform.data.table = function (`_data`, ...)
 # basically transform.data.frame with data.table instead of data.frame, and retains key
 {
   if (!cedta()) return(NextMethod()) # nocov
-  e = eval(substitute(list(...)), `_data`, parent.frame())
-  tags = names(e)
-  inx = chmatch(tags, names(`_data`))
-  matched = !is.na(inx)
-  if (any(matched)) {
-    .Call(C_unlock, `_data`) # fix for #1641, now covered by test 104.2
-    `_data`[,inx[matched]] = e[matched]
-    `_data` = as.data.table(`_data`)
-  }
-  if (!all(matched)) {
-    ans = do.call("data.table", c(list(`_data`), e[!matched]))
-  } else {
-    ans = `_data`
-  }
-  key.cols = key(`_data`)
-  if (!any(tags %chin% key.cols)) {
-    setattr(ans, "sorted", key.cols)
-  }
-  ans
+  `_data` = copy(`_data`)
```

Comment:
full copy() is to be conservative, but a .shallow() copy should also work.

the conservatism is because transform() returns a copy by default, and existing behavior does so, so we should respect that. but if .shallow() can be used it should be faster (please play around a bit to make sure the behavior stays the same, I haven't played with .shallow in quite some time)

Reply:
I can't reproduce Ofek's twice-as-slow performance. On the original sample of 1e7 rows, data.table seems 4% slower than base, which seems reasonable. I believe we should keep the `copy` to maintain the existing behavior.

``` r
library(data.table)
library(microbenchmark)

df <- data.frame(x = runif(n = 1e7))
dt <- as.data.table(df)
microbenchmark(transform(df, y = round(x)),transform(dt, y = round(x)), times = 10)
#> Unit: milliseconds
#>                         expr      min       lq     mean   median       uq
#>  transform(df, y = round(x)) 865.6029 884.9914 944.9678 896.1184 1034.661
#>  transform(dt, y = round(x)) 884.6268 913.6477 968.5802 932.4312 1058.328
#>       max neval
#>  1040.406    10
#>  1105.619    10
```